### DISCUSSION TITLE
Leetcode plugin now working

### DISCUSSION THREAD

**Discussion by cloydlau** [lowlighter/metrics#1261] [Q&A] [answered]
Hello, the new leetcode plugin seems not working, am I missing something?

![image](https://user-images.githubusercontent.com/31238760/196329781-b7c7fd27-b7d7-4df2-a5e5-7d9c02d14200.png)

And I wonder does this plugin support `leetcode.cn`? Or is a PR accepetable?

**Comment by lowlighter** [comment, marked as answer]
Hi 👋 !

I enabled the plugin on the server so you should be able to use the `leetcode` plugin now 👍 

These kind of errors occurs when a plugin is not enabled on the server's `settings.json`. 
For [metrics.lecoq.io](https://metrics.lecoq.io), usually most token-base and api/cpu intensive plugins are disabled, but in this case it was just an oversight since I forgot to update the config file (the leetcode api doesn't require any token so it's fine to let anyone use this plugin).

____

As for `leetcode.cn` it is not supported currently, but if the GraphQL API is exactly the same as `leetcode.com` then it should be quite easy to implement and a PR would be welcome

One would needs to update [plugins/leetcode/metadata.yml](https://github.com/lowlighter/metrics/blob/bddd67e46cf1d5fbeb4b37597a7e37b02fc5e14b/source/plugins/leetcode/metadata.yml) to add:
```yml
 plugin_leetcode_url:
    description: |
      LeetCode URL
    type: string
    default: https://leetcode.com
```

And patch this line to extract `url` too (no need to typecheck and default it, the `imports.metadata.plugins.leetcode.inputs` will automatically do it from the provided `metadata.yml`) :
https://github.com/lowlighter/metrics/blob/bddd67e46cf1d5fbeb4b37597a7e37b02fc5e14b/source/plugins/leetcode/index.mjs#L10

Replace all occurences of `"https://leetcode.com/graphql"` by `` `${url}/graphql` ``

And users should be able to pass `leetcode.url=https://leetcode.cn` to change the endpoint

**Comment by cloydlau** [comment]
Thanks for the PR guide, it's very helpful! I followed the contribution guide and kind of stucked. I'm not familiar with the docker stuff. A `docker: invalid reference format` show up when I typing `docker run --rm --entrypoint="" -p=127.0.0.1:80:3000 --volume=/settings.json:/metrics/settings.json ghcr.io/lowlighter/metrics:latest npm start`.

**Reply by lowlighter** [reply]
It could be the mounted volume that errors since `/settings.json` may not be the correct path on your local machine, or the port option (I'm not sure if ip binding is supported, but just `-p=80:3000` should do it). Also if you start the container as it, it'll have the code from `@latest` branch since you would need to mount your local repository to `/metrics` too (using `--volume=/path/to/repo:/metrics`)

Usually it's simpler to set up a [local server for development](https://github.com/lowlighter/metrics/blob/master/.github/readme/partials/documentation/setup/local.md) rather than using the docker container (which is more suited for deployments), but it may take a bit of time (I'll check to make this process a bit easier and faster in the future)

To be honest, it may be a bit overkill for this small feature addition and you can probably directly open the PR if you feel confident with your code.
The CI should catch any breaking errors (note it's a bit long though), and the code review should catch any logic errors so it's probably fine if you don't test it yourself (though I won't stop you if you absolutly want to test it 😅)

To give you an idea, the PR should look similar to https://github.com/lowlighter/metrics/pull/1082/files (though obviously it wasn't intended for the same feature):
- a new option in metadata.yml
- the option extracted from `imports.metadata.plugins`
- small code editions from index.mjs to handle the extracted option
